framework/Optimizers/stepManipulators/ConjugateGradient.py

```diff
@@ -0,0 +1,436 @@
+  ##########################
+  @classmethod
+  def getInputSpecification(cls):
+    """
+      Method to get a reference to a class that specifies the input data for class cls.
+      @ In, cls, the class for which we are retrieving the specification
+      @ Out, inputSpecification, InputData.ParameterInput, class to use for specifying input of cls.
+    """
+    specs = super(ConjugateGradient, cls).getInputSpecification()
+    return specs
+
+  @classmethod
+  def getSolutionExportVariableNames(cls):
+    """
+      Compiles a list of acceptable SolutionExport variable options.
+      @ In, None
+      @ Out, vars, list(str), list of acceptable variable names
+    """
+    ok = super(ConjugateGradient, cls).getSolutionExportVariableNames()
+    ok['CG_task'] = 'for ConjugateGradient, current task of line search. FD suggests continuing the search, and CONV indicates the line search converged and will pivot.'
+    return ok
+
+  def __init__(self):
+    """
+      Constructor.
+      @ In, None
+      @ Out, None
+    """
+    StepManipulator.__init__(self)
+    ## Instance Variable Initialization
+    # public
+    self.needsAccessToAcceptance = True # if True, then this stepManip may need to modify opt point acceptance criteria
+    # _protected
+    self._persistence = None     # consecutive line search converges until acceptance
+    # __private
+    # additional methods
+
+  def handleInput(self, specs):
+    """
+      Read input specs
+      @ In, specs, InputData.ParameterInput, parameter specs interpreted
+      @ Out, None
+    """
+    #specs = specs
+    growth = specs.findFirst('growthFactor')
+    if growth is not None:
+      self._growth = growth.value
+    shrink = specs.findFirst('shrinkFactor')
+    if shrink is not None:
+      self._shrink = shrink.value
+
+  def initialize(self, optVars, persistence=None, **kwargs):
+    """
+      initializes this object
+      @ In, optVars, list(str), list of optimization variables (e.g. input space)
+      @ In, persistence, integer, successive converges required to consider total convergence
+      @ In, kwargs, dict, additional unused arguments
+      @ Out, None
+    """
+    StepManipulator.initialize(self, optVars, **kwargs)
+    self._persistence = persistence
+
+
+  ###############
+  # Run Methods #
+  ###############
+  def initialStepSize(self, numOptVars=None, scaling=0.05, **kwargs):
+    """
+      Provides an initial step size
+      @ In, numOptVars, int, number of optimization variables
+      @ In, scaling, float, optional, scaling factor
+    """
+    return mathUtils.hyperdiagonal(np.ones(numOptVars) * scaling)
+
+  def step(self, prevOpt, gradientHist=None, prevStepSize=None, objVar=None, **kwargs):
+    """
+      calculates the step size and direction to take
+      @ In, prevOpt, dict, previous opt point
+      @ In, gradientHist, deque, list of gradient dictionaries with 0 being oldest; versors
+      @ In, prevStepSize, deque, list of float step sizes
```

Comment:
specify the optional and what "happens" when they are not inputted

Reply:
ok